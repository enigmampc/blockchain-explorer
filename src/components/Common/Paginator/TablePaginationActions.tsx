import React from 'react'
import IconButton from '@material-ui/core/IconButton'
import FirstPageIcon from '@material-ui/icons/FirstPage'
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft'
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight'
import LastPageIcon from '@material-ui/icons/LastPage'

interface TablePaginationActionsProps extends React.HTMLAttributes<HTMLDivElement> {
  count: number
  page: number
  rowsPerPage: number
  onChangePage: CallableFunction
}

const TablePaginationActions = ({ count, page, rowsPerPage, onChangePage }: TablePaginationActionsProps) => {
  const handleFirstPageButtonClick = (event: React.MouseEvent<HTMLButtonElement> | null) => {
    onChangePage(event, 0)
  }

  const handleBackButtonClick = (event: React.MouseEvent<HTMLButtonElement> | null) => {
    onChangePage(event, page - 1)
  }

  const handleNextButtonClick = (event: React.MouseEvent<HTMLButtonElement> | null) => {
    onChangePage(event, page + 1)
  }

  const handleLastPageButtonClick = (event: React.MouseEvent<HTMLButtonElement> | null) => {
    onChangePage(event, Math.max(0, Math.ceil(count / rowsPerPage) - 1))
  }

  return (
    <div style={{ flexShrink: 0 }}>
      <IconButton onClick={handleFirstPageButtonClick} disabled={page === 0} aria-label="first page">
        <FirstPageIcon />
      </IconButton>
      <IconButton onClick={handleBackButtonClick} disabled={page === 0} aria-label="previous page">
        <KeyboardArrowLeft />
      </IconButton>
      <IconButton
        aria-label="next page"
        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        onClick={handleNextButtonClick}
      >
        <KeyboardArrowRight />
      </IconButton>
      <IconButton
        onClick={handleLastPageButtonClick}
        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        aria-label="last page"
      >
        <LastPageIcon />
      </IconButton>
    </div>
  )
}

export default TablePaginationActions
