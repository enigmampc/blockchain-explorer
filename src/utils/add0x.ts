const add0x = (hex: string): string => (hex ? (hex.indexOf('0x') === 0 ? hex : `0x${hex}`) : '')

export default add0x
