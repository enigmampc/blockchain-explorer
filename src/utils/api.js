import { formatNumber } from './numeralFormats'

export const cryptoCompare = async () => {
  const url = new URL('https://min-api.cryptocompare.com/data/pricemultifull')
  const params = {
    fsyms: 'ENG',
    tsyms: 'USD',
  }
  url.search = new URLSearchParams(params).toString()
  const {
    RAW: {
      ENG: {
        USD: { PRICE, MKTCAP },
      },
    },
  } = await (await fetch(url)).json().catch(console.error)

  return {
    price: formatNumber(PRICE, 4),
    marketCap: formatNumber(MKTCAP, 0),
  }
}

export const coinMarketCap = async () => {
  const url = new URL('https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest')
  const params = {
    id: 2044,
  }
  url.search = new URLSearchParams(params).toString()
  const headers = new Headers({
    'X-CMC_PRO_API_KEY': process.env.REACT_APP_COINMARKET_CAP_API_KEY,
  })
  const init = {
    method: 'GET',
    headers,
    cache: 'default',
  }
  const request = new Request(url, init)
  const response = await (await fetch(request)).json()

  if (response.error_code) throw new Error(response.error_message)

  const {
    data: {
      [params.id]: {
        quote: {
          USD: { market_cap, price },
        },
      },
    },
  } = response

  return {
    price: formatNumber(price, 4),
    marketCap: formatNumber(market_cap, 0),
  }
}

export const binance = async () => {
  const binanceURL = 'https://api.binance.com/api/v1/ticker/price?symbol='
  const { price } = await (await fetch(`${binanceURL}ENGBTC`)).json().catch()

  return {
    price: formatNumber(price, 4),
    marketCap: '',
  }
}
