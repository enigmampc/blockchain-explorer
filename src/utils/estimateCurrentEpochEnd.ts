import ethApi from './eth'
import { EpochProps } from '../components/Epochs'

const calculateRangesDurations = (range: number[]) => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = (range: number[]) => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

export interface EstimatesCurrentEpochEnd {
  pendingTime: number
  finishBlockNumber: number
}

const estimateCurrentEpochEnd = async (epochs?: EpochProps[], average = false): Promise<EstimatesCurrentEpochEnd> => {
  let pendingTime = 0 // estimated life time for the current epoch
  let finishBlockNumber = 0 // estimated block for when the current epoch may end

  if (!epochs) return { pendingTime, finishBlockNumber }

  const sortedEpochs = epochs.sort((a, b) => +b.id - +a.id)
  const currentEpoch = sortedEpochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 40, 22]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = sortedEpochs
    .reduce((acc: number[], epoch, index) => {
      if (index !== 0) {
        acc.push(+epoch.startBlockNumber)
        acc.push(+epoch.endBlockNumber)
      }

      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])
    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)
    finishBlockNumber = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlockNumber - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount
    pendingTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { pendingTime, finishBlockNumber }
}

export default estimateCurrentEpochEnd
