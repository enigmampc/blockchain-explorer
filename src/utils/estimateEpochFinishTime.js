import ethApi from './eth'

const calculateRangesDurations = range => {
  const acc = []

  for (let i = 0; i < range.length; i += 2) {
    const rangeElement = range[i]
    acc.push(rangeElement - range[i + 1])
  }

  return acc
}

const distanceMedian = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (length === 0) return length

  durations.sort((a, b) => a - b)

  const half = Math.floor(length / 2)

  return length % 2 ? durations[half] : Math.floor((durations[half - 1] + durations[half]) / 2)
}

const distanceAverage = range => {
  const durations = calculateRangesDurations(range)
  const length = durations.length

  if (!length) return 0

  const sum = durations.reduce((a, b) => a + b)
  return Math.floor(sum / length)
}

const estimateEpochFinishTimes = async (epochs = [], average = false) => {
  let finishTime = 0 // when the current epoch may end
  let finishBlock = 0 // where the current epoch may end
  const currentEpoch = epochs.slice(0, 1)[0] // currently active epoch

  // builds an array of the form [10, 1, 21, 11, 22, 40]
  // where 10 is the last block of the first epoch in the collection
  // and 1 is the first block of the first epoch in the collection
  // 21 is the last block of the second epoch and 11 is it's first block and so on
  const blocksRange = epochs
    .reduce((acc, epoch, index) => {
      const current = index === 0
      if (!current) {
        acc.push(+epoch.startBlockNumber) // start block number for current epoch
        acc.push(+epochs[index - 1].startBlockNumber - 1) // finish block number for current epoch
      }
      return acc
    }, [])
    .reverse()

  try {
    const [currentBlock, epochsRangesTimestamps] = await Promise.all([
      ethApi.getBlockNumber(),
      ethApi.getBatchBlocksTimestamps(blocksRange),
    ])

    const epochBlockCount = average ? distanceAverage(blocksRange) : distanceMedian(blocksRange)
    const epochDuration = average ? distanceAverage(epochsRangesTimestamps) : distanceMedian(epochsRangesTimestamps)

    finishBlock = +currentEpoch.startBlockNumber + epochBlockCount
    const currentBlocksLeft = finishBlock - +currentBlock
    const currentBlocksTimes = epochDuration / epochBlockCount

    finishTime = Math.floor(currentBlocksLeft * currentBlocksTimes) * 1000 // milliseconds
  } catch (e) {
    console.error('Failed to retrieve blocks tiemstamps', e)
  }

  return { finishTime, finishBlock }
}

export default estimateEpochFinishTimes
