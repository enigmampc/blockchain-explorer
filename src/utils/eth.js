import Web3 from 'web3'
// import EnigmaContract from './Enigma.json'
import getNetworkDetailsBy from './networks'

export const ENG_DECIMALS = 8

class EthAPI {
  constructor() {
    if (process.env.REACT_APP_ETH_NETWORK_ID) {
      const { url } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)
      this.web3 = new Web3(new Web3.providers.HttpProvider(url))
    } else {
      this.web3 = new Web3(new Web3.providers.HttpProvider(process.env.REACT_APP_ETH_URL))
    }
  }

  getBlockNumber() {
    return this.web3.eth.getBlockNumber()
  }

  async getBlockTimestamp(blockNumber) {
    const block = await this.web3.eth.getBlock(blockNumber)
    return (block && block.timestamp) || '0'
  }

  async getBatchBlocksTimestamps(range) {
    const batch = new this.web3.BatchRequest()

    const blocks = range.map(blockNumber => {
      return new Promise((resolve, reject) => {
        const request = this.web3.eth.getBlock.request(blockNumber, (error, data) => {
          if (error) {
            reject(error)
          } else {
            resolve(data)
          }
        })
        batch.add(request)
      })
    })

    batch.execute()

    return (await Promise.all(blocks)).map(({ timestamp }) => timestamp)
  }
}

export default new EthAPI()
