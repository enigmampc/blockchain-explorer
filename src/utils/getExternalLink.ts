import getNetworkDetailsBy from './networks'

const { name: networkName } = getNetworkDetailsBy('id')(process.env.REACT_APP_ETH_NETWORK_ID)

const getExternalLink = (ethContract: string) => {
  const subdomain = ['mainnet', 'local', 'unknown'].includes(networkName) ? '' : `${networkName}.`
  return `https://${subdomain}${process.env.REACT_APP_ETHERSCAN_URL}/${ethContract}`
}

export default getExternalLink
