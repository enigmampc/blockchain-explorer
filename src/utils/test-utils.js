import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import theme from '../theme'

const getStyledComponents = StyledComponent => {
  const componentClass = StyledComponent().type.styledComponentId
  return document.getElementsByClassName(componentClass)
}

const history = createBrowserHistory()

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <Router history={history}>{children}</Router>
  </ThemeProvider>
)

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'

export { customRender as render, getStyledComponents, history }
