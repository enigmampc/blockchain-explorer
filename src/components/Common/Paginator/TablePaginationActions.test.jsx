import React from 'react'
import { render, fireEvent } from '../../../utils/test-utils'
import TablePaginationActions from './TablePaginationActions'

it(`should render TablePaginationActions without crashing`, () => {
  // Given
  const count = 60
  const page = 0
  const rowsPerPage = 10
  const onChangePage = jest.fn()

  // When
  const { container, getAllByRole } = render(
    <TablePaginationActions count={count} page={page} rowsPerPage={rowsPerPage} onChangePage={onChangePage} />,
  )
  const buttons = getAllByRole('button')

  // Then
  expect(buttons[0]).toHaveClass('Mui-disabled')
  expect(buttons[1]).toHaveClass('Mui-disabled')
  expect(container).toMatchSnapshot()
})

it(`should only call callback when action is available starting with first page`, () => {
  // Given
  const count = 60
  const page = 0
  const rowsPerPage = 10
  const onChangePage = jest.fn()

  // When
  const { getAllByRole } = render(
    <TablePaginationActions count={count} page={page} rowsPerPage={rowsPerPage} onChangePage={onChangePage} />,
  )
  const buttons = getAllByRole('button')

  // Then
  fireEvent.click(buttons[2])
  expect(onChangePage).toHaveBeenCalledWith(1)
  fireEvent.click(buttons[3])
  expect(onChangePage).toHaveBeenCalledWith(5)
})

it(`should only call callback when action is available starting with last page`, () => {
  // Given
  const count = 60
  const page = 5
  const rowsPerPage = 10
  const onChangePage = jest.fn()

  // When
  const { getAllByRole } = render(
    <TablePaginationActions count={count} page={page} rowsPerPage={rowsPerPage} onChangePage={onChangePage} />,
  )
  const buttons = getAllByRole('button')

  // Then
  fireEvent.click(buttons[1])
  expect(onChangePage).toHaveBeenCalledWith(4)
  fireEvent.click(buttons[0])
  expect(onChangePage).toHaveBeenCalledWith(0)
})
