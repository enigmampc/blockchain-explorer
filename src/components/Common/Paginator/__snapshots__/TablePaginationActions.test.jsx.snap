// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TablePaginationActions without crashing 1`] = `
<div>
  <div
    style="flex-shrink: 0;"
  >
    <button
      aria-label="first page"
      class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
          />
          <path
            d="M24 24H0V0h24v24z"
            fill="none"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="previous page"
      class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="next page"
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-label="last page"
      class="MuiButtonBase-root MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
